### ISSUE TITLE
[Validator] [htmlparser.js] Lack of end tags is possibly misleading

### ISSUE THREAD

**Issue by sidkshatriya** [ampproject/amphtml#3713]
In various tests across the validator, one comes across various amp tags which don't have end tags.

e.g.

https://github.com/ampproject/amphtml/blob/569fb55a9c9f0724c06155543282f766780472fe/validator/testdata/feature_tests/mandatory_dimensions.html#L114-L120 
```html
<!-- Layout of responsive/fixed without width/height -->
    <amp-img src="img" layout="responsive">
    <amp-img src="img" layout="fixed">
    <amp-img src="img" layout="responsive" width="42">
    <amp-img src="img" layout="fixed" height="42">
    <amp-img src="img" layout="fixed" height="auto">
    <amp-img src="img" layout="fixed" width="auto" height="42">
```
Notice the lack of `</amp-img>`

https://github.com/ampproject/amphtml/blob/de471567c924ce51e401248ef69c001cee599cfc/validator/testdata/feature_tests/regexps.html#L75-L79
```html
  <amp-audio src="https://exmaple.com/audio" layout="fixed" autoplay="">
  <amp-audio src="https://example.com/audio" layout="fixed" autoplay="desktop">
  <amp-audio src="https://example.com/audio" layout="fixed" autoplay="invalid">
  <amp-audio src="https://example.com/audio" layout="fixed" autoplay="desktopfoo">
  <amp-audio src="https://example.com/audio" layout="fixed" autoplay="foodesktop">
```

Notice the lack of `</amp-audio>`

**The indentation of the HTML seems to indicate that an end closing tag is implicitly understood before the next tag starts. However that is not the case, sadly.**

So:
```html
    <amp-img src="img" layout="responsive">
    <amp-img src="img" layout="fixed">
    <amp-img src="img" layout="responsive" width="42">
    <amp-img src="img" layout="fixed" height="42">
    <amp-img src="img" layout="fixed" height="auto">
    <amp-img src="img" layout="fixed" width="auto" height="42">
```

is **not** 
```html
    <amp-img src="img" layout="responsive"></amp-img>
    <amp-img src="img" layout="fixed"></amp-img>
    <amp-img src="img" layout="responsive" width="42"></amp-img>
    <amp-img src="img" layout="fixed" height="42"></amp-img>
    <amp-img src="img" layout="fixed" height="auto"></amp-img>
    <amp-img src="img" layout="fixed" width="auto" height="42"></amp-img>
```

But it is actually:
```html
<amp-img src="img" layout="responsive">
    <amp-img src="img" layout="fixed">
        <amp-img src="img" layout="responsive" width="42">
            <amp-img src="img" layout="fixed" height="42">
                <amp-img src="img" layout="fixed" height="auto">
                    <amp-img src="img" layout="fixed" width="auto" height="42">
                    </amp-img>
                </amp-img>
            </amp-img>
        </amp-img>
    </amp-img>
</amp-img>
```

This can be proved by added a simple test to `htmlparser_test.js`

```javascript
  it('multiple images', () => {
    const handler = new LoggingHandler();
    const parser = new amp.htmlparser.HtmlParser();
    parser.parse(handler, '<amp-img src="one.gif"><amp-img src="two.gif">');
    expect(handler.log).toEqual([
      'startDoc()', 'startTag(BODY,[])',
      'startTag(IMG,[src,one.gif])', 'endTag(img)','startTag(IMG,[src,two.gif])', 'endTag(IMG)',
      'endTag(BODY)', 'endDoc()'
    ]);
  });
```
```
Failures:
1) HtmlParser multiple images
  Message:
    Expected [ 'startDoc()', 'startTag(BODY,[])', 'startTag(AMP-IMG,[src,one.gif])', 'startTag(AMP-IMG,
[src,two.gif])', 'endTag(AMP-IMG)', 'endTag(AMP-IMG)', 'endTag(BODY)', 'endDoc()' ] to equal [
'startDoc()', 'startTag(BODY,[])', 'startTag(IMG,[src,one.gif])', 'endTag(img)', 'startTag(IMG,[src,two.gif])',
'endTag(IMG)', 'endTag(BODY)', 'endDoc()' ].
```

The above output shows that htmlparser nests the tags in absence of a closing tag. I wonder if that is the intention of the tests (`mandatory_dimensions.html` and `regexps.html`) to nest tags in this fashion?

Some time ago I filed a related ticket https://github.com/ampproject/amphtml/issues/2551 . The resolution on the ticket seemed to state that this was a HTML validation issue and the browser would insert the end tags automatically. Yes, I agree the browser will insert end tags automatically but it is likely to nest `<amp-img>` and `<amp-audio>` (which is not likely to have been the intention of the test). In fact `htmlparser.js` does the same thing.

So I guess in conclusion:
- If your intention is to nest tags in this way in the tests, perhaps the indentation of the tags should be changed in the test or otherwise its is confusing -- one assumes that the end tag will be provided before the next tag. Alternatively please add a comment, perhaps?
- If your intention is not to nest the tags this way then you should add end tags in the test, perhaps?

The [amp php library](https://github.com/Lullabot/amp-library) is a ported subset of your validator so its necessary to go into these details to make sure our implementation is doing similar things as yours. 

Incidently the HTML parser we use for our validator automatically provides end tags in such a way that the tags nest (This is reassuring to us because the `htmlparser.js` seems to do the same).

**Comment by powdercloud** [comment]
Hi Sidharth, I think your diagnosis as to how the code behaves is correct, as in, htmlparser.js emits a tag stream that nests the tags. Further, it is good that the amp php library agrees with that. Thank you.
I think the tag nesting behavior and so on is or should be tested with htmlparser_test.js - just like you did. If you see some gap, I think that'd be a good contribution as it ties down the behavior and that is useful since with HTML5 I think parsers are supposed to have specific behaviors when they see less than optimal HTML.
About the feature tests for the validator, they don't care much about this particular thing. They try to exercise the validator - which, yes, runs htmlparser.js to get a tag stream - but the main point here is to test the validator behavior, which at the moment is not dependent on this nesting vs sibling issue for amp-img. We're not sending these tests through the W3C HTML5 validator, and we're not extremely careful about the indentation, we just put the stuff into there that we need to get reasonable validator coverage and some comments to make it readable / maintainable. The comments focus on the validator logic, not so much the html parsing logic.
So, acknowledging what's up with the nesting amp-img tags that you pointed out, what do you suggest we do? Is htmlparser_test.js inadequate as is? Do we need to go through and add closing tags in our feature test, or perhaps closing tags plus missing ones with comments or similar? Or some dedicated feature test for this? I'm not sure. I could see us just letting it be for now and add to the tests as needed when there's some problem with a document that should or should not validate. Thanks for your analysis and your thoughts on this.

**Comment by sidkshatriya** [comment]
@powdercloud Thanks for your detailed response.

I'm somewhat conflicted about the future (suggested) course of action. The main thing I was worried about is our html parser nesting the tags. Since `htmlparser.js` does the same and you've confirmed that it works as intended I think we can leave it at that. I also tried an experiment on Google Chrome without the ending tags and there too I can confirm that the `<amp-img>` tags nest.

So something like
```html
<amp-img src="img1.jpg"><amp-img src="img2.jpg>
```

In the browser becomes something like:
```html
<amp-img src="img1.jpg>
    <amp-img layout="responsive" src="img2.jgp>
       <img src="img2.jpg>
    </amp-img>
    <img src="img2.jpg">
</amp-img>
```

The display looks messed up but I think this is correct behavior because the user has not provided closing tags. (This behavior of nesting of course is not specific to `<amp-img>` but should hold true for any other tags like `<amp-audio>` mentioned above -- I just didn't try it out in the browser). I don't think we should be alarmed about the messed up display -- people forget to close their `<div>` tags and things nest in funny ways too -- so the onus must be on the user to make sure they provided valid html.

So I would suggest two alternatives (both are fine):
- Leave things as they are, aware that the parser behaves as it does.
- Close the tags in the tests. I think you're not trying to test the nesting -- its just happening by chance so you might as well close them. I for instance saw this syntax and assumed that the tags will be closed automatically before the next tag starts (but this was not the case when I looked deeper). By adding closing tags you might ensure that people who see these tests don't think that this is a valid way to write your tags.

Both alternatives are logical and I leave it to you folks. My main goal, which was clarification, has been achieved. Thanks again for spending the time on this.
